src/app/qgsgeometryvalidationservice.cpp

```diff
@@ -0,0 +1,454 @@
+
+    // Tell all checks to stop asap
+    const auto feedbacks = mLayerChecks[layer].topologyCheckFeedbacks;
+    for ( QgsFeedback *feedback : feedbacks )
+    {
+      if ( feedback )
+        feedback->cancel();
+    }
+
+    futureWatcher->waitForFinished();
+    mLayerChecks[layer].topologyCheckFutureWatcher = nullptr;
+  }
+}
+
+void QgsGeometryValidationService::clearTopologyChecks( QgsVectorLayer *layer )
+{
+  QList<std::shared_ptr<QgsGeometryCheckError>> &allErrors = mLayerChecks[layer].topologyCheckErrors;
+  allErrors.clear();
+
+  emit topologyChecksCleared( layer );
+}
+
+void QgsGeometryValidationService::invalidateTopologyChecks( QgsVectorLayer *layer )
+{
+  cancelTopologyCheck( layer );
+  layer->setAllowCommit( false );
+}
+
+void QgsGeometryValidationService::processFeature( QgsVectorLayer *layer, QgsFeatureId fid )
+{
+  emit geometryCheckStarted( layer, fid );
+
+  QgsFeature feature = layer->getFeature( fid );
+
+  const auto &checks = mLayerChecks[layer].singleFeatureChecks;
+  mLayerChecks[layer].singleFeatureCheckErrors.remove( fid );
+
+  // The errors are going to be sent out via a signal. We cannot keep ownership in here (or can we?)
+  // nor can we be sure that a single slot is connected to the signal. So make it a shared_ptr.
+  QList<std::shared_ptr<QgsSingleGeometryCheckError>> allErrors;
+  for ( QgsSingleGeometryCheck *check : checks )
+  {
+    const auto errors = check->processGeometry( feature.geometry() );
+
+    for ( auto error : errors )
+      allErrors.append( std::shared_ptr<QgsSingleGeometryCheckError>( error ) );
+  }
+
+  if ( !allErrors.empty() )
+    mLayerChecks[layer].singleFeatureCheckErrors.insert( fid, allErrors );
+
+  emit geometryCheckCompleted( layer, fid, allErrors );
+}
+
+void QgsGeometryValidationService::setMessageBar( QgsMessageBar *messageBar )
+{
+  mMessageBar = messageBar;
+}
+
+void QgsGeometryValidationService::triggerTopologyChecks( QgsVectorLayer *layer )
+{
+  cancelTopologyCheck( layer );
+  clearTopologyChecks( layer );
+
+  QgsFeatureIds affectedFeatureIds;
+  if ( layer->editBuffer() )
+  {
+    affectedFeatureIds = layer->editBuffer()->changedGeometries().keys().toSet();
+    affectedFeatureIds.unite( layer->editBuffer()->addedFeatures().keys().toSet() );
+  }
+
+  QgsFeaturePool *featurePool = mFeaturePools.value( layer->id() );
+  if ( !featurePool )
+  {
+    featurePool = new QgsVectorLayerFeaturePool( layer );
+    mFeaturePools.insert( layer->id(), featurePool );
+  }
+
+  QList<std::shared_ptr<QgsGeometryCheckError>> &allErrors = mLayerChecks[layer].topologyCheckErrors;
+
+  QMap<QString, QgsFeatureIds> layerIds;
+
+  QgsFeatureRequest request = QgsFeatureRequest( affectedFeatureIds ).setSubsetOfAttributes( QgsAttributeList() );
+  QgsFeatureIterator it = layer->getFeatures( request );
+  QgsFeature feature;
+  QgsRectangle area;
+  while ( it.nextFeature( feature ) )
+  {
+    area.combineExtentWith( feature.geometry().boundingBox() );
```

Comment:
Random thought -- maybe we need a feature request flag for "bounding boxes only". It'd certainly speed up a lot of code like this.

Reply:
Random thought response: I'd like to have "derived values" in requests, like "request specific virtual fields".

```
request.setFlags(QgsFeatureRequest::NoGeometry)
request.addDerivedValue("bbox", "bounds($geometry)")
```

This would be very flexible, although I admit for this particular use case a custom flag could be justified